### PR TITLE
Treat decompiled sources separately from metadata sources

### LINKED ISSUE #25244: Turning of decompilation doesn't take effect until you restart VS
1. Turn on 
image

2. _Go to definition_ for an external API
3. Close document window
4. Turn off above option
5. _Go to definition_ for the same API

**Expected:** For metadata as source to be used
**Actual:** The code is decompiled

### LINKED ISSUE #25255: "Enable Navigation to Decompiled Sources" option only partially works in a VB.NET project
After enabling the &quot;Enable Navigation to Decompiled Sources&quot; option I attempted to Go To Definition of `.Show()` of a `System.Windows.Forms.Form` from my VB.NET project. This incorrectly opens just the decompiled file rather than also jumping to the function definition. However, it seems to work perfectly fine from a C# project, additionally I can jump to the definition of the class itself from a VB project.

Also, it seems to be syntax highlighting the code in the decompiled file with VB syntax rather than C# syntax.

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/207892/enable-navigation-to-decompiled-sources-option-onl.html
VSTS ticketId: 576951_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

### LINKED ISSUE #25731: Using Peek Definition disabled decompilation for that type
**Version Used**: Visual Studio 2017.7 Preview 3 [27526.1.d15.7]

**Steps to Reproduce**:

1.  `Peek Definition` on a type defined in metadata
2.  Notice that the metadata type is not decompiled in to source
3.  Use `Go To Definition`

**Expected Behavior**:
Navigated to decompiled sources

**Actual Behavior**:
Navigated to metadata as source

### LINKED ISSUE #26329: Decompilation from a VB projet produces a C# file, but then opens it in a VB workspace.
For example:

image

This clearly breaks everything since the C# code can't be well understood by the VB LS.  

Options:

1. Use a VB decompiler in VB settings.
2. If we can only use a C# decompiler, then make a C# project to put hte code into.  Don't fake up VB project for that code.

### LINKED ISSUE #27187: [Disassembly] body of method not shown after answering No and then Yes
**Version Used**: 15.8 Preview 1.1

🔗 Copied from internal bug [600249](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/600249).

### Steps to reproduce:

1. Create a C# Class Library (.NET Framework) with a method:

    ```csharp
    public static int AbsoluteValue(int input)
    {
        var result = Math.Abs(input);
        return result;
    }
    ```

2. Build the library 
3. Close the current solution and create a new C# Class Library (.NET Framework) 
4. Add a reference to the DLL created in step 2 
5. Add a call to the public static method defined in step 1 
6. Place the caret on the call, and press F12 to navigate to the definition
7. Click No
8. Close opened decompiled library
9. Press F12 to navigate to the definition again
10. Click Yes
 
### Expected:

The AbsoluteValue method appears, and the body is shown with a call to Math.Abs

### Actual:

The AbsoluteValue method appears, but the body is not visible.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #25244
Fixes #25255
Fixes #25731
Fixes #26329
Fixes #27187

<details><summary>Ask Mode template not completed</summary>

### Customer scenario

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

### Bugs this fixes

(either VSO or GitHub links)

### Workarounds, if any

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

### Risk

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

### Performance impact

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

### Root cause analysis

How did we miss it?  What tests are we adding to guard against it in the future?

### How was the bug found?

(E.g. customer reported it vs. ad hoc testing)

### Test documentation updated?

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by sharwell** [comment]
retest windows_debug_vs-integration_prtest please

**Comment by sharwell** [comment]
retest windows_release_vs-integration_prtest please

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 1]
```diff
@@ -63,16 +64,17 @@ private static string CreateMutexName(string directoryName)
...
+        private string GetRootPathWithGuid_NoLock(bool allowDecompilation)
         {
-            if (_rootTemporaryPathWithGuid == null)
+            ref string rootTemporaryPathWithGuid = ref (allowDecompilation ? ref _rootDecompiledTemporaryPathWithGuid : ref _rootMetadataTemporaryPathWithGuid);
+            if (rootTemporaryPathWithGuid == null)
             {
                 var guidString = Guid.NewGuid().ToString("N");
-                _rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
+                rootTemporaryPathWithGuid = Path.Combine(_rootTemporaryPath, guidString);
                 _mutex = new Mutex(initiallyOwned: true, name: CreateMutexName(guidString));
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: This now may be overwriting a previous Mutex created by our process; I'm guessing if the Mutex GCs then another process might clean up things under us? I believe the right fix here is to preserve a single folder, and just maybe add another layer of folders inside of it, i.e. MetadataAsSource-{guid}\Decompiled vs. MetadataAsSource-{guid}\Metadata. Or, hold onto two mutexes, but I think going back to a single directory is easier, and also scales if/when we create a third type of something to go in here...

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs [hunk 2]
```diff
@@ -21,16 +22,24 @@ internal sealed class MetadataAsSourceGeneratedFileInfo
...
         private readonly ParseOptions _parseOptions;
 
-        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType)
+        public MetadataAsSourceGeneratedFileInfo(string rootPath, Project sourceProject, INamedTypeSymbol topLevelNamedType, bool allowDecompilation)
         {
             this.SourceProjectId = sourceProject.Id;
-            _parseOptions = sourceProject.ParseOptions;
             this.Workspace = sourceProject.Solution.Workspace;
-            this.LanguageName = sourceProject.Language;
+            if (allowDecompilation && sourceProject.Language != LanguageNames.CSharp)
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: If you just move this after this.LanguageName =, then I think you can reduce this all back to _parseOptions = Workspace.Services.GetLanguageServices(this.LanguageName).GetRequriedService().ParseOptions;. Basically, you've already figured out once that if it's decompiled it's always C#; no reason to do it twice.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: Two mutexes is easy with the current pattern. Yay ref locals!

**Comment by sharwell** [comment]
@jasonmalinowski Updated based on review feedback. Planning to rebase this back to 2 commits once the review completes.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs [hunk 3]
```diff
@@ -63,16 +70,18 @@ private static string CreateMutexName(string directoryName)
             return "MetadataAsSource-" + directoryName;
         }
 
-        private string GetRootPathWithGuid_NoLock()
+        private string GetRootPathWithGuid_NoLock(bool allowDecompilation)
         {
-            if (_rootTemporaryPathWithGuid == null)
+            ref string rootTemporaryPathWithGuid = ref (allowDecompilation ? ref _rootDecompiledTemporaryPathWithGuid : ref _rootMetadataTemporaryPathWithGuid);
+            ref Mutex mutex = ref (allowDecompilation ? ref _decompiledMutex : ref _metadataMutex);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: i personally hate this :)  but i guess i will have to grow to live with these sorts of patterns :)

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs [hunk 4]
```diff
@@ -21,16 +22,24 @@ internal sealed class MetadataAsSourceGeneratedFileInfo
...
-            this.LanguageName = sourceProject.Language;
+            this.LanguageName = allowDecompilation ? LanguageNames.CSharp : sourceProject.Language;
+            if (sourceProject.Language == LanguageName)
+            {
+                _parseOptions = sourceProject.ParseOptions;
+            }
+            else
+            {
+                _parseOptions = Workspace.Services.GetLanguageServices(LanguageName).GetRequiredService<ISyntaxTreeFactoryService>().GetDefaultParseOptionsWithLatestLanguageVersion();
+            }
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: nit: consider ?:

Comment (sharwell) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: Opted against it here. The two branches seem different enough that I like to keep it obvious.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs (hunk 3 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs]: Yeah, I admire your desire to play with new things, but if we just went back to a single mutex/folder and generated the paths differently, it's just two lines with no fancy tricks whatsoever.

"If you write your code as cleverly as possible..."

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: Why still having the branches at all? When would the source project's parse options be different than the default parse options for the language in an observable way?

I'm not saying there's not a reason for this....I just can't tell what it is. I can't read your mind (not sure if good thing or bad thing?) but do we need a comment here?

**Comment by sharwell** [comment]
@jasonmalinowski main feedback should be addressed now.

File: src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs [hunk 5]
```diff
@@ -21,16 +22,24 @@ internal sealed class MetadataAsSourceGeneratedFileInfo
...
+
             this.References = sourceProject.MetadataReferences.ToImmutableArray();
             this.AssemblyIdentity = topLevelNamedType.ContainingAssembly.Identity;
 
-            var extension = sourceProject.Language == LanguageNames.CSharp ? ".cs" : ".vb";
+            var extension = LanguageName == LanguageNames.CSharp ? ".cs" : ".vb";
 
             var directoryName = Guid.NewGuid().ToString("N");
             this.TemporaryFilePath = Path.Combine(rootPath, directoryName, topLevelNamedType.Name + extension);
```
Comment (jasonmalinowski) [inline src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: Doesn't this still need to have some differentiation for the decompiled vs. not?

Comment (sharwell) [reply src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceGeneratedFileInfo.cs]: It does. Two levels of directory names are used when the keys don't match; this one will still get one unique GUID in the path where before it was getting two.

**Comment by jasonmalinowski** [comment]
@jinujoseph I see this is assigned to me now, who is doing the work to take this for QB approval?

**Comment by jinujoseph** [comment]
approved to merge for 15.8.Preview5

**Pull request merged**